homeassistant/components/media_player/kef.py

```diff
@@ -0,0 +1,302 @@
+SUPPORT_KEF_CLIENT_DEVICE = (
+    SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP | SUPPORT_VOLUME_MUTE |
+    SUPPORT_SELECT_SOURCE | SUPPORT_TURN_OFF
+)
+
+CONF_TURN_ON_SERVICE = 'turn_on_service'
+CONF_TURN_ON_DATA = 'turn_on_data'
+
+# yaml configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_TURN_ON_SERVICE): cv.service,
+    vol.Optional(CONF_TURN_ON_DATA): cv.string,
+})
+
+
+# setup of component
+def setup_platform(hass, config, add_devices,
+                   discovery_info=None):
+    """Set up Kef platform."""
+    import pykef
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    name = config.get(CONF_NAME)
+    speaker = pykef.KefSpeaker(host, port)
+    turn_on_service = config.get(CONF_TURN_ON_SERVICE)
+    turn_on_data = config.get(CONF_TURN_ON_DATA)
+
+    # configure source options to communicate to HA
+    sources = collections.OrderedDict([
+        ('1', str(pykef.InputSource.Wifi)),
+        ('2', str(pykef.InputSource.Bluetooth)),
+        ('3', str(pykef.InputSource.Aux)),
+        ('4', str(pykef.InputSource.Opt)),
+        ('5', str(pykef.InputSource.Usb))
+    ])
+
+    _LOGGER.debug(
+        "Setting up %s, using host: %s, port: %s, name: %s",
+        DATA_KEF, str(host), str(port), str(name))
+    _LOGGER.debug("Setting up source_dict %s", str(sources))
+
+    # Add devices
+    add_devices([KefClientDevice(
+        name, speaker, turn_on_service, turn_on_data, sources, hass
+    )])
+
+
+class States(Enum):
+    """States for the a KefClientDevice."""
+
+    Online = 1
+    Offline = 2
+    TurningOn = 3
+    TurningOff = 4
+
+
+class KefClientDevice(MediaPlayerDevice):
+    """Kef Player Object."""
+
+    def __init__(self, name, speaker, turn_on_service, turn_on_data,
+                 source_dict, hass):
+        """Initialize the media player."""
+        self._hass = hass
+        self._name = name
+        self._source_dict = source_dict
+        self._speaker = speaker
+        self._turn_on_service = turn_on_service
+        self._turn_on_data = turn_on_data
+
+        # set internal state to None
+        self._state = None
+        self._mute = None
+        self._source = None
+        self._volume = None
+        self._update_timeout = time.time() - CHANGE_STATE_TIMEOUT
+
+    def __wait_for_online_state(self):
+        """Use this function to wait for online state."""
+        time_to_wait = WAIT_FOR_ONLINE_STATE
+        while time_to_wait > 0 and self._state is not States.Online:
+            time_to_sleep = 0.1
+            time_to_wait -= time_to_sleep
+            time.sleep(time_to_sleep)
+            if self._state is States.TurningOn:
+                time_to_wait = 10
+
+    def __internal_state(self):
+        """Return text with the internal states, just for debugging."""
```

Comment:
Remove this. Debugging should be done in the used module.

Reply:
Ack

Reply:
Ack